Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm

```diff
@@ -254,6 +254,57 @@
     }
 }
 
+void WebExtensionController::didCommitLoadForFrame(WebPageProxyIdentifier pageID, WebCore::FrameIdentifier frameID, URL frameURL)
+{
+    WebExtensionContext::EventListenerTypeSet listenerTypes;
+    listenerTypes.add(WebExtensionEventListenerType::WebNavigationOnCommitted);
+    // FIXME: Is there a more precise time to fire the DOMContentLoaded event? This matches Safari's behavior.
+    listenerTypes.add(WebExtensionEventListenerType::WebNavigationOnDOMContentLoaded);
+
+    for (auto& context : m_extensionContexts) {
+        // FIXME: We need to turn pageID into a _WKWebExtensionTab and pass that here.
+        if (!context->hasPermission(frameURL))
+            continue;
+
+        context->fireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
+            context->sendToProcessesForEvent(WebExtensionEventListenerType::WebNavigationOnCommitted, Messages::WebExtensionContextProxy::DispatchWebNavigationOnCommittedEvent(pageID, frameID, frameURL));
+            context->sendToProcessesForEvent(WebExtensionEventListenerType::WebNavigationOnDOMContentLoaded, Messages::WebExtensionContextProxy::DispatchWebNavigationOnDOMContentLoadedEvent(pageID, frameID, frameURL));
+        });
+    }
+}
+
+void WebExtensionController::didFinishLoadForFrame(WebPageProxyIdentifier pageID, WebCore::FrameIdentifier frameID, URL frameURL)
+{
+    WebExtensionContext::EventListenerTypeSet listenerTypes;
+    listenerTypes.add(WebExtensionEventListenerType::WebNavigationOnCompleted);
+
+    for (auto& context : m_extensionContexts) {
+        // FIXME: We need to turn pageID into a _WKWebExtensionTab and pass that here.
+        if (!context->hasPermission(frameURL))
+            continue;
+
+        context->fireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
```

Comment:
Might be able to just do: `fireEvents({ WebExtensionEventListenerType::WebNavigationOnCompleted }, ...`

Reply:
That would be cool! I'll try that.

Reply:
It looks like in that case I'd use the event type twice.

I'll do the auto suggestion and the inline definition you suggested before.